### PR TITLE
Fix for "Bus error (core dumped)" when saving recipe state

### PR DESCRIPTION
**Pull request by calvinpelletier**
#### Context
Quick fix for this issue: https://github.com/pytorch/torchtune/issues/1018

An alternative solution is to save recipe states with the epoch appended to the filename. This might be a better solution but there are trade-offs. There is some overlap between this issue and https://github.com/pytorch/torchtune/issues/988 so I'll make an RFC discussing some potential changes to how we do checkpointing. 

#### Changelog
- disable mmap when loading recipe state

#### Test plan

Run finetune then interrupt after 1 epoch: 
`tune run lora_finetune_single_device --config llama3/8B_lora_single_device seed=0 epochs=4 max_steps_per_epoch=2 tokenizer.path=/home/cpelletier/model/llama3-8b/original/tokenizer.model checkpointer.checkpoint_dir=/home/cpelletier/model/llama3-8b/original/ checkpointer.output_dir=/home/cpelletier/out/bus_error/`

Resume finetune: `tune run lora_finetune_single_device --config llama3/8B_lora_single_device seed=0 epochs=4 max_steps_per_epoch=2 tokenizer.path=/home/cpelletier/model/llama3-8b/original/tokenizer.model checkpointer.checkpoint_dir=/home/cpelletier/out/bus_error/ checkpointer.adapter_checkpoint=adapter_0.pt checkpointer.recipe_checkpoint=recipe_state.pt resume_from_checkpoint=True checkpointer.output_dir=/home/cpelletier/out/bus_error/`

Before the change, this resulted in `Bus error (core dumped)` when saving `recipe_state.pt` after resuming.

### CODE REVIEW

**Comment by joecummings** [comment]
Can you take a look at updating our test to catch this kind of failure?

File: torchtune/utils/_checkpointing/_checkpointer_utils.py [hunk 1]
```diff
@@ -53,7 +53,9 @@ def get_path(input_dir: Path, filename: str, missing_ok: bool = False) -> Path:
     return file_path
 
 
-def safe_torch_load(checkpoint_path: Path, weights_only: bool = True) -> Dict[str, Any]:
+def safe_torch_load(
+    checkpoint_path: Path, weights_only: bool = True, mmap: bool = True
```
Comment (kartikayk) [inline torchtune/utils/_checkpointing/_checkpointer_utils.py:57]: not related to this PR, but @ebsmothers we've come back to the API of torch.load with this change. At this point I don't think there's any use of safe_torch_load other than as a single wrapper for safetensors vs not. We probably should think about refactoring this in a clean way.

**Review by kartikayk — APPROVED** [review]
Thanks for a) catching this, b) debugging this and c) the fix!

**Review by ebsmothers — APPROVED** [review]

File: recipes/lora_finetune_single_device.py [hunk 2]
```diff
@@ -156,20 +156,18 @@ def _update_recipe_state(self, ckpt_dict: Dict[str, Any]) -> None:
...
             or self.max_steps_per_epoch != ckpt_dict[utils.MAX_STEPS_KEY]
         ):
             warn(
-                message="""Configured value for seed, epochs or max_steps_per_epoch
+                message="""Configured value for seed or max_steps_per_epoch
                 does not match the value stored in checkpoint."""
             )
         self.seed = utils.set_seed(seed=ckpt_dict[utils.SEED_KEY])
         self.epochs_run = ckpt_dict[utils.EPOCHS_KEY]
-        self.total_epochs = ckpt_dict[utils.TOTAL_EPOCHS_KEY]
```
Comment (calvinpelletier) [inline recipes/lora_finetune_single_device.py:172]: @joecummings I made a quick fix to the `test_training_state_on_resume` test so that we'll notice errors like this, but it required slightly different configurations for the pre- and post-resume runs (first `epochs=2` and then `epochs=3`), which was not allowed by this `_update_recipe_state` above, so i removed the overwrite for total epochs. 

Is there a reason for this overwrite? IMO we should allow users to resume a training job with a different number of total epochs, like for example if they finish a training job and then decide they want to continue training for a few more epochs. To support this, we should also save the recipe state during the final checkpoint.

Thoughts? Should I remove the epoch overwrite for the other recipes? Or should I find a different way to test for this error?

Comment (joecummings) [reply recipes/lora_finetune_single_device.py:172]: I 100% agree with this approach, but maybe we should opt to be overly communicative here? Log a warning if the total epochs from the ckpt_dict differ from the total epochs requested?
